Introduce blog post for disk-based k-NN (#3616)
opensearch-project/project-website: _posts/2025-02-03-Reduce-Cost-with-Disk-based-Vector-Search.md

We chose four different cluster configurations to test against and used [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk) and OpenSearch 2.18 to provision the clusters. In general, we chose the cluster and index configuration to follow production recommendations. For instance, we configured replica shards and dedicated cluster manager nodes. In addition, we targeted the shard count to have between 2 and 3 vCPUs per shard.

| Name        | Data node count | Data node type | Data node disk size | Data node disk type | JVM Size | Primary shard count | replica shards | compression level |
|-------------|-----------------|----------------|---------------------|---------------------|----------|---------------------|----------------|-------------------|
| in_memory   | 8               | r6g.8xlarge    | 300                 | EBS                 | 32       | 40                  | 1              | 1x                |
| on_disk_8x  | 10              | r6gd.2xlarge   | 474                 | Instance            | 32       | 15                  | 1              | 8x                |

> | on_disk_16x | 6               | r6gd.2xlarge   | 474                 | Instance            | 32  9    | 1                   | 16x            |
> | on_disk_32x | 4               | r6gd.2xlarge   | 474                 | Instance            | 32  6    | 1                   | 32x            |

Reviewer: ```suggestion
| on_disk_16x | 6               | r6gd.2xlarge   | 474                 | Instance            | 32 | 9    | 1                   | 16x            |
| on_disk_32x | 4               | r6gd.2xlarge   | 474                 | Instance            | 32 | 6    | 1                   | 32x            |
```

In order to run the tests, we also made a few other optimizations to tune performance:

1. Retrieved ids from doc values in order to reduce fetch time
2. Disabled source storage 

The test procedure was fairly simple. All tests were executed with OSB. See the [vectorsearch workloads](https://github.com/opensearch-project/opensearch-benchmark-workloads/tree/main/vectorsearch) for more details.  We did the following:

1. Ingest the data set
2. Force merged to 5 segments per shard
3. Warmed the index up
4. Ran the Single-client latency test
5. Ran the Multi-client throughput tests
6. Repeated 4 and 5 with disk-based re-score setting disabled to show what the results would be without re-scoring, effectively searching over a compressed index without re-scoring.

#### Results

From these experiments, we obtained the following results:

| Metric/Configuration              | in-memory | on_disk_8x | in_memory_8x | on_disk_16x | in_memory_16x | on_disk_32x | in_memory_32x |
|-----------------------------------|-----------|------------|--------------|-------------|---------------|-------------|---------------|

> | recall@100 (ratio)                | 0.95      | 0.98       | 0.98         | 0.97        | 0.96          | 0.94        | 0.95          |

Reviewer: in memory recall for 32x compression is better than on_disk? any reason why?

> | recall@100 (ratio)                | 0.95      | 0.98       | 0.98         | 0.97        | 0.96          | 0.94        | 0.95          |

Reviewer: This shows that 32x compression just works and we should not add on_disk here

Author: Added both just for sake of transparency. For some data sets, the re-scoring does not significantly help.

---
layout: post
title:  "Reduce costs with disk-based vector search"
authors:
   - jmazane
   - vikash
   - vamshin
   - kolchfa

> date: 2025-02-03

Reviewer 2: Update blog publish date to:
2025-02-19

has_science_table: true
categories:
  - technical-posts
meta_keywords: Disk-based vector search in OpenSearch, quantization, ANN search, compression, performance testing, binary quantization

> meta_description: Learn how to reduce your OpenSearch costs for vector search using the new disk-based vector search.

Reviewer 2: Please update the meta with the following: 

_meta_keywords:_ disk based vector search, vector search, vector quantization techniques, OpenSearch 2.17, binary quantization, natural language embedding models

_meta_description:_ Learn how OpenSearch significantly reduces vector search costs using disk-based methods. Discover quantization techniques and practical implementation strategies for efficient large-scale vector search operations.

However, quantization comes with a trade-off: The more compression applied, the greater the potential loss in search accuracy. Fortunately, with improvements in secondary storage, it's possible to trade some search latency for accurate nearest-neighbor search in a low-memory environment.

## Disk-based vector search in OpenSearch

In 2.17, we introduced an `on_disk` mode for `knn_vector` fields that enables disk-based vector search in OpenSearch, allowing searches to run in much lower memory environments. This is achieved using a two-phase query approach, illustrated in the following diagram.  

![High-level two-phase query approach](/assets/media/blog-images/2025-02-03-Reduce-Cost-with-Disk-based-Vector-Search/hla.png){:class="img-centered"}

First, during ingestion, we build indexes using configurable quantization mechanisms and store the full-precision vectors on disk. The quantized indexes, which reside fully in memory during search, require significantly less memory due to compression.  

> During search, we first query the quantized indexes to retrieve more than k candidate results. Then we lazily load the full-precision vectors of these candidates into memory and recompute the distance to the query. This reordering step ensures accuracy, and we return only the top K results.  

Reviewer 3: ```suggestion
During search, we first query the quantized indexes to retrieve more than k candidate results. Then we lazily load the full-precision vectors of these candidates into memory and recompute the distance to the query. This reordering step ensures accuracy, and we return only the top k results.  
```

Empirically, this two-phase approach has demonstrated the ability to produce [high-recall results](https://github.com/opensearch-project/k-NN/issues/1779#user-content-appendix-b-baseline-rescore-experiments) across a variety of quantization techniques.

### Introducing online binary quantization  

As part of this feature, we also expanded our quantization support. Previously, achieving ≥8x compression in OpenSearch required using product quantization. While effective, product quantization requires a training step before ingestion can begin and requires users to manage their models.  

> **In 2.17, we introduced several new online quantization techniques for 8x, 16x, and 32x compression—no pretraining needed. You can begin ingestion immediately without additional model preparation.** This simplifies the process and reduces overhead. Take a look at this technical deep dive to learn how we developed these capabilities.

Reviewer 3: Right below lines 55--91

| Name              | Space type | Normalized | Dimension | Index vector count | Notes                                                                                                                                       |
|-------------------|--------------|----------------|-----------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------|
| `sift`              | `l2`           | No             | 128       | 1,000,000          | Classic scale-invariant feature transform (SIFT) image descriptor dataset.                                                                                                       |
| `minillm-msmarco `  | `l2`           | Yes            | 384       | 1,000,000          | Used [`minillm`](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2 ) to encode a sample of `ms-marco`.                              |
| `mpnet-msmarco`     | `l2`           | Yes            | 768       | 1,000,000          | Used [`mpnet`](https://huggingface.co/sentence-transformers/all-mpnet-base-v2) to encode a sample of `ms-marco`.                                |

> | `mxbai-msmarco`     | `cosine`       | Yes            | 1024      | 1,000,000          | Used [`mxbai`](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1 ) to encode a sample of `ms-marco`.                                    |

Reviewer 4: ```suggestion
| `mxbai-msmarco`     | `cosine`       | Yes            | 1,024      | 1,000,000          | Used [`mxbai`](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1 ) to encode a sample of `ms-marco`.                                    |
```

| `tasb-msmarco`      | `ip`           | No             | 768       | 1,000,000          | Used [`tasb`](https://huggingface.co/sentence-transformers/msmarco-distilbert-base-tas-b) to encode a sample of `ms-marco`.                    |
| `e5small-msmarco`   | `l2`           | Yes            | 384       | 8,841,823          | Used [`e5small`](https://huggingface.co/intfloat/e5-small ) to encode a larger sample of `ms-marco`.                                 |
| `sNowflake-msmarco` | `l2`           | Yes            | 768       | 8,841,823          | Used [`snowflake-arctic-embed-m`](https://huggingface.co/Snowflake/snowflake-arctic-embed-m) to encode a larger sample of `ms-marco`. |
| `clip-flickr`       | `l2`           | No             | 512       | 6,637,685          | Used the `clip` model to encode a random sample of images from Flickr.                                                                         |

> We conducted experiments comparing `in_memory` and `on_disk` modes using both `16x` and `32x` compression levels. Key testing parameters included:

Author: ```suggestion
We conducted experiments comparing `in_memory` and `on_disk` modes using `1x`, `2x`, `4x`, `8x`, `16x` and `32x` compression levels. Key testing parameters included:
```

> We conducted experiments comparing `in_memory` and `on_disk` modes using `1x`, `2x`, `4x`, `8x`, `16x` and `32x` compression levels. Key testing parameters included:

Reviewer 4: ```suggestion
We conducted experiments comparing `in_memory` and `on_disk` modes using `1x`, `2x`, `4x`, `8x`, `16x`, and `32x` compression levels. Key testing parameters included the following:
```

- `on_disk` mode: Includes a rescoring stage

> - `in_memory` mode: No rescoring stage

Reviewer 4: ```suggestion
- `in_memory` mode: Does not include a rescoring stage
```

- Hardware configuration:

>   - `in_memory` tests: AWS EC2 r6g.4xlarge instances

Reviewer 4: ```suggestion
  - `in_memory` tests: Amazon Elastic Compute Cloud (Amazon EC2) r6g.4xlarge instances
```

>   - `on_disk` tests: AWS EC2 r6gd.4xlarge instances with attached SSD (as opposed to EBS storage)

Reviewer 4: ```suggestion
  - `on_disk` tests: Amazon EC2 r6gd.4xlarge instances with an attached solid-state drive (SSD)---as opposed to Amazon Elastic Block Store (Amazon EBS) storage
```

- Resource controls: Used Docker to manage OpenSearch's memory and CPU allocation

We obtained the following results, comparing recall to compression level.

![Million Vector Results](/assets/media/blog-images/2025-02-03-Reduce-Cost-with-Disk-based-Vector-Search/mill-results.png){:class="img-centered"}

> The results show that combining binary quantization with full-precision rescoring can significantly improve recall while maintaining reasonable latency in low-memory environments. However, performance varies by dataset. For example, while this approach improves the `sift` dataset performance, the recall remains relatively low. We recommend testing with your specific dataset to determine the optimal configuration.

Author: ```suggestion
The results show that combining quantization with full-precision rescoring can significantly improve recall while maintaining reasonable latency in low-memory environments. However, performance varies by dataset. For example, while this approach improves the `sift` dataset performance, the recall remains relatively low. We recommend testing with your specific dataset to determine the optimal configuration.
```

> The results show that combining quantization with full-precision rescoring can significantly improve recall while maintaining reasonable latency in low-memory environments. However, performance varies by dataset. For example, while this approach improves the `sift` dataset performance, the recall remains relatively low. We recommend testing with your specific dataset to determine the optimal configuration.

Reviewer 4: ```suggestion
The results show that combining quantization with full-precision rescoring can significantly improve recall while maintaining reasonable latency in low-memory environments. However, performance varies by dataset. For example, while this approach improves the performance of the `sift` dataset, the recall remains relatively low. We recommend testing using your specific dataset to determine the optimal configuration.
```

### Large-scale tests

> 

Reviewer 4: Line 211: I assume we're going to fill in the link brackets?

> In addition to the smaller scale tests, we also ran a test on a larger dataset, [](https://huggingface.co/datasets/Cohere/msmarco-v2.1-embed-english-v3). We set up several different tests to showcase how this disk-based vector search compares to `in_memory` vector search at `8x`, `16x`, and `32x` compression levels. We ran these tests using the following configuration.

Reviewer 4: ```suggestion
In addition to the smaller-scale tests, we also ran a test on a larger dataset, [](https://huggingface.co/datasets/Cohere/msmarco-v2.1-embed-english-v3). We set up several different tests to showcase how this disk-based vector search compares to `in_memory` vector search at `8x`, `16x`, and `32x` compression levels. We ran these tests using the following configuration.
```

> | Dataset      | [MS Marco 2.1 encoded with the Cohere v3 model](https://huggingface.co/datasets/Cohere/msmarco-v2.1-embed-english-v3)                                       |

Reviewer 4: ```suggestion
| Dataset      | [MS MARCO 2.1 encoded with the Cohere v3 model](https://huggingface.co/datasets/Cohere/msmarco-v2.1-embed-english-v3)                                       |
```

> | Dimension  | 1024                                                                     |

Reviewer 4: ```suggestion
| Dimension  | 1,024                                                                     |
```

| Normalized      | Yes                                                                      |
| Space type | Cosine (inner product over normalized data)                              |
| Index vectors | 113M                                                                     |

> We tested four different cluster configurations using [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk) with OpenSearch 2.18. We selected the cluster and index configuration to follow production recommendations. For example, we configured replica shards and dedicated cluster manager nodes. In addition, we targeted a shard count that provides 2 to 3 vCPUs per shard. The following table presents the configurations for these tests.

Reviewer 4: Should "the" precede the link?

Author: yes good catch

> We tested four different cluster configurations using [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk) with OpenSearch 2.18. We selected the cluster and index configuration to follow production recommendations. For example, we configured replica shards and dedicated cluster manager nodes. In addition, we targeted a shard count that provides 2 to 3 vCPUs per shard. The following table presents the configurations for these tests.

Reviewer 3: ```suggestion
We tested four different cluster configurations using the [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk) with OpenSearch 2.18. We selected the cluster and index configuration to follow production recommendations. For example, we configured replica shards and dedicated cluster manager nodes. In addition, we targeted a shard count that provides 2 to 3 vCPUs per shard. The following table presents the configurations for these tests.
```

| Name        | Data node count | Data node type | Data node disk size | Data node disk type | JVM size | Primary shard count | Replica shards | Compression level |
|-------------|-----------------|----------------|---------------------|---------------------|----------|---------------------|----------------|-------------------|
| `in_memory`   | 8               | r6g.8xlarge    | 300                 | EBS                 | 32       | 40                  | 1              | 1x                |
| `on_disk_8x`  | 10              | r6gd.2xlarge   | 474                 | Instance            | 32       | 15                  | 1              | 8x                |
| `on_disk_16x` | 6               | r6gd.2xlarge   | 474                 | Instance            | 32 | 9    | 1                   | 16x            |
| `on_disk_32x` | 4               | r6gd.2xlarge   | 474                 | Instance            | 32 | 6    | 1                   | 32x            |

> In the table, note that clusters with higher compression levels use significantly less resources than those without compression.

Reviewer 4: ```suggestion
In the table, note that clusters with higher compression levels use significantly fewer resources than those without compression.
```

> To optimize performance for our tests, we made these additional adjustments:

Reviewer 4: ```suggestion
To optimize performance in our tests, we made these additional adjustments:
```

- Retrieved IDs from `doc_values` to reduce fetch time
- Disabled `_source` storage 

We ran all tests using OpenSearch Benchmark [vector search workloads](https://github.com/opensearch-project/opensearch-benchmark-workloads/tree/main/vectorsearch), following this procedure:

1. Ingest the dataset.
2. Force merge the index to 5 segments per shard.
3. Run warm-up queries to load the index into memory.
4. Run the single-client latency test.

> 5. Run the multi-client throughput tests.

Reviewer 4: Confirm that singular "test" on line 240 but plural "tests" on line 241 is correct.

Author: should both be test

> 6. Repeat steps 4 and 5 with disk-based rescore disabled to measure performance on a compressed index without rescoring.

Reviewer 4: ```suggestion
6. Repeat steps 4 and 5 with disk-based rescoring disabled to measure performance on a compressed index without rescoring.
```

> 5. Run the multi-client throughput tests.

Reviewer 3: ```suggestion
5. Run the multi-client throughput test.
```

We obtained the following large-scale test results.

> | Metric/Configuration              | `in-memory` | `on_disk_8x` | `in_memory_8x` | `on_disk_16x` | `in_memory_16x` | `on_disk_32x` | `in_memory_32x` |

Reviewer 4: ```suggestion
| Metric/configuration              | `in-memory` | `on_disk_8x` | `in_memory_8x` | `on_disk_16x` | `in_memory_16x` | `on_disk_32x` | `in_memory_32x` |
```

| 1-client mean throughput (QPS)           | 40.64     | 11.03      | 42.33        | 10.06       | 43.95         | 10.65       | 31.83         |
| 4-client p90 search latency (ms)  | 25.82     | 97.19      | 20.40        | 220.19      | 18.09         | 244.52      | 46.13         |
| 4-client mean throughput (QPS)           | 162.80    | 45.86      | 204.62       | 25.80       | 193.27        | 25.28       | 146.77        |
| 8-client p90 search latency (ms) | 27.69     | 95.05      | 30.88        | 414.20      | 27.94         | 429.79      | 25.07         |
| 8-client mean throughput (QPS)           | 306.70    | 95.22      | 305.86       | 26.34       | 343.95        | 25.75       | 376.60        |

Interestingly, for this dataset, the on-disk approach with rescoring produces similar recall to the in-memory approach without rescoring, but the in-memory approach is substantially faster. This is most likely because the Cohere v3 model has been optimized to work very well with binary quantized data (see [this blog post](https://cohere.com/blog/int8-binary-embeddings)).

## Learnings

> 

Reviewer 4: Line 260: We haven't actually referenced ANN prior to this. Instead of "ANN approach", do we mean "nearest neighbor approach"?

> Our testing shows that the two-phase ANN approach performs effectively in low-memory environments, though results vary significantly by dataset. When running your own experiments, we recommend testing with both enabled and disabled `index.knn.disk.vector.shard_level_rescoring_disabled` setting to measure the performance benefit for your use case. Additionally, with disk-based search, ensure that the secondary storage is optimized for high read traffic---we found that SSDs generally provide the best results.

Reviewer 4: ```suggestion
Our testing shows that the two-phase ANN approach performs effectively in low-memory environments, though results vary significantly by dataset. When running your own experiments, we recommend testing with the `index.knn.disk.vector.shard_level_rescoring_disabled` setting both enabled and disabled to measure the performance benefit for your use case. Additionally, with disk-based search, ensure that your secondary storage is optimized for high read traffic---we found that SSDs generally provide the best results.
```

## What's next?

> We have many new and exciting features coming for vector search in OpenSearch. In the future releases, we'll focus on improving quantization performance for all datasets, eliminating the need for fine-tuning. Follow our [GitHub repo](https://github.com/opensearch-project/k-NN) for continued improvements in both performance and functionality. As always, we welcome and appreciate your contributions and feature requests!

Reviewer 4: ```suggestion
We have many new and exciting features coming for vector search in OpenSearch. In future versions, we'll focus on improving quantization performance for all datasets, eliminating the need for fine-tuning. Follow our [GitHub repo](https://github.com/opensearch-project/k-NN) for continued improvements in both performance and functionality. As always, we welcome and appreciate your contributions and feature requests!
```
